keps/sig-node/0014-runtime-class.md

```diff
@@ -0,0 +1,359 @@
+
+### Goals
+
+- Provide a mechanism for surfacing container runtime properties to the control plane
+- Support multiple runtimes per-cluster, and provide a mechanism for users to select the desired
+  runtime
+
+### Non-Goals
+
+- RuntimeClass is NOT a general policy mechanism
+- RuntimeClass is NOT "NodeClass". Although different nodes may run different runtimes, in general
+  RuntimeClass should not be a cross product of runtime properties and node properties.
+
+The following goals are out-of-scope for the initial implementation, but may be explored in a future
+iteration:
+
+- Automatic runtime or feature discovery - initially RuntimeClasses are manually defined (by the
+  cluster admin or provider), and are asserted to be an accurate representation of the runtime.
+- Scheduling in heterogeneous clusters - it is possible to operate a heterogeneous cluster
+  (different runtime configurations on different nodes) through scheduling primitives like
+  `NodeAffinity` and `Taints+Tolerations`, but the user is responsible for setting these up and
+  automatic runtime-aware scheduling is out-of-scope.
+- Define standardized or conformant runtime classes - although I would like to declare some
+  predefined RuntimeClasses with specific properties, doing so is out-of-scope for this initial KEP.
+- [Pod Overhead][] - Although RuntimeClass is likely to be the configuration mechanism of choice,
+  the details of how pod resource overhead will be implemented is out of scope for this KEP.
+- Provide a mechanism to dynamically register or provision additional runtimes.
+- Requiring specific RuntimeClasses according to policy. This should be addressed by other
+  cluster-level policy mechanisms, such as PodSecurityPolicy.
+- "Fitting" a RuntimeClass to pod requirements - In other words, specifying runtime properties and
+  letting the system match an appropriate RuntimeClass, rather than explicitly assigning a
+  RuntimeClass by name. This approach can increase portability, but can be added seamlessly in a
+  future iteration.
+
+[Pod Overhead]: https://docs.google.com/document/d/1EJKT4gyl58-kzt2bnwkv08MIUZ6lkDpXcxkHqCvvAp4/edit
+
+### User Stories
+
+- As a cluster operator, I want to provide multiple runtime options to support a wide variety of
+  workloads. Examples include native linux containers, "sandboxed" containers, and windows
+  containers.
+- As a cluster operator, I want to provide stable rolling upgrades of nodes.
+- As an application developer, I want to select the runtime that best fits my workload.
+- As an application developer, I don't want to study the nitty-gritty details of different runtime
+  implementations.
+- As an application developer, I want my application to be portable across clusters that use similar
+  but different variants of a "class" of runtimes.
+- As an application operator, I want to know precisely how the application is running, and which
+  features are being used.
```

Comment:
you mean auditing applications by finding the runtimeclass and looking at what features of the runtime those class enable ?

Reply:
See the second paragraph of the motivation section. Just because a feature is specified in the podspec doesn't mean the pod is running with that feature.

Reply:
ok so we are saying that by specifying the runtimeclass, the control plane knows what features that class supports, so i can match what is specified in the podspec with what is supported in the class to surface errors. May be this needs to be written somewhere or its obvious  to everyone

Reply:
@yujuhong yes, that's the idea. Right now that's not the case (e.g. selinux), but with "supported features" enforcement, you could assume that if you specified something in the PodSpec it's actually being used.